semrel.go

```diff
@@ -211,84 +214,117 @@ func (listRefsQuery) vars(repo *Repository, perPage int, cursor string) map[stri
+	if prerelease != "" {
+		// locate pre-release
+		eg.Go(func() error {
+			log.Println("Searching for pre-release version:", prerelease)
+			n := 0
+			for r, err := range repo.tags(prerelease) {
+				// this will often find some false positives that we need to skip over
+				if err != nil {
+					return err
+				} else if rPreRel, _, _ := strings.Cut(r.Version.Prerelease(), "."); rPreRel == prerelease {
+					log.Println("Found latest matching pre-release version:", r.Version.String())
+					lastPreRelease = r
+					break
 				}
+				n++
 			}
-		}
+			log.Println("Skipped", n, "non-matching pre-release versions")
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
 
-		if !q.Repository.Refs.PageInfo.HasNextPage {
-			break
+	if prerelease == "" {
+		return lastMainRelease, nil
+	}
+
+	if lastPreRelease != nil {
+		// if we found both a pre-release and default branch release tag, we need to
+		// see which is newer
+		if lastMainRelease == nil {
+			return lastPreRelease, nil
+		} else if lastPreRelease.Version.GreaterThan(lastMainRelease.Version) {
+			log.Println("Last pre-release version is higher than last main release")
+			return lastPreRelease, nil
 		}
-		cursor = string(q.Repository.Refs.PageInfo.EndCursor)
 	}
+	return lastMainRelease, nil
+}
 
-	// pretend the latest release is from verRange, applying any pre-release tag it has
-	if verRangeVer, err := semver.NewVersion(verRange); err != nil {
-		return nil, fmt.Errorf("unable to parse maintained version %q as semver: %w", verRange, err)
-	} else if _, verRangePre, vrHasPre := strings.Cut(verRange, "-"); !vrHasPre {
-		// verRange is just a version, no pre-release segment. Attach its version to
-		// the latest release's commit
-		return &Release{lastRelease.SHA, verRangeVer}, nil
-	} else if verWithPreRel, err := verRangeVer.SetPrerelease(verRangePre); err != nil {
-		return nil, fmt.Errorf("failed to initialize pre-release for mainted version %q: %w", verRange, err)
-	} else {
-		// verRange is a version and a pre-release tag. Attach this combo to the
-		// latest release's commit
-		return &Release{lastRelease.SHA, &verWithPreRel}, nil
+// tags generates a sequence of Release objects built from tags that match the
+// given query string (which may be empty) in reverse chronological order by
+// commit date.
+func (repo *Repository) tags(query string) iter.Seq2[*Release, error] {
+	return func(yield func(*Release, error) bool) {
+		var cursor string
+	PAGES:
+		for {
+			var q listRefsQuery
+			err := repo.GQLClient.Query(repo.Ctx, &q, q.vars(repo, query, 100, cursor))
+			if err != nil {
+				yield(nil, err)
+				break PAGES
+			}
+		NODES:
+			for _, n := range q.Repository.Refs.Nodes {
+				version, err := semver.NewVersion(strings.TrimPrefix(string(n.Name), "refs/tags/"))
+				if err != nil {
+					// silently ignore non-semver tags
+					continue NODES
+				}
+				r := &Release{string(n.Target.Oid), version}
+				if !yield(r, nil) {
```

Comment:
What causes yield to fail here?  

The internal gotos in here are a little confusing.  A comment for what part of the processing the gotos are skipping might be nice.

Reply:
This is new go iterator/generator stuff. The compiler generates the glue code such that `yield` returns `false` here if the calling `for` loop does a `break` or `return` or the like.

Ref: https://go.dev/blog/range-functions#range-over-some-function-types

Reply:
> internal gotos

Huh? there's no `goto` in here, just continue/break using labels so it can escape multiple layers of for loop

Reply:
Would it be clearer if `break PAGES` were just `return`? I guess `continue PAGES` is redundant, since that's only ever from within that inner loop in the first place

Reply:
Oh yeah, I guess I misunderstood how break LABEL works in go.. yeah, a `break PAGES` as a return would be more obvious to me.. but that's probably just because it's more C like..